src/node/lattice.rs

```diff
@@ -0,0 +1,518 @@
+                    // If the event we are trying to add is dependent on the DFS node, then
+                    // add an edge from the DFS node to the event to be added.
+                    if event < *node || event == *node {
+                        // Check if any of the children of the DFS node should be dependent on the
+                        // event we are trying to add.
+                        let mut remove_edges: Vec<NodeIndex<u32>> = Vec::new();
+                        for child in forest.neighbors(nx) {
+                            let child_node: &OperatorEvent =
+                                &forest.node_weight(child).unwrap().as_ref().unwrap();
+                            if *child_node < event {
+                                // The child has a dependency on the node being added. 
+                                // Break the dependency from the DFS node to its child, and add a
+                                // dependency from the node to be added to the child.
+                                add_edges_to.push(child);
+                                remove_edges.push(child);
+                            }
+                        }
+
+                        // Add a dependency from the DFS node to the node being added, only if the
+                        // partial order says so.
+                        if event < *node {
+                            add_edges_from.push(nx);
+                            for child in &remove_edges {
+                                let edge = forest.find_edge(nx, *child).unwrap();
+                                forest.remove_edge(edge);
+                            }
+                            dependencies.insert(nx);
+                        }
+                    } else {
+                        // The current event had no dependent children, but the node in DFS is
+                        // dependent on the current event. Usually, this is resolved in a level
+                        // above, but add edges if the node is root.
+                        if forest.neighbors_directed(nx, Direction::Incoming).count() == 0 {
+                            add_edges_to.push(nx);
+                            for (i, n) in run_queue.iter().enumerate() {
+                                if n.index() == nx.index() {
+                                    remove_roots.push((i, nx));
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    // Reached a node that is already executing, but hasn't been completed.
+                    // The current node will probably get added as a root. Add dependencies to the
+                    // children, if any.
+                    for child in forest.neighbors(nx) {
+                        let child_node = forest.node_weight(child).unwrap().as_ref().unwrap();
+                        if *child_node < event {
```

Comment:
If you don't want to deref:
```suggestion
                        if child_node < &event {
```